incuna_auth/middleware/permission.py

```diff
@@ -0,0 +1,232 @@
+from django.contrib import messages
+from django.http import HttpResponseForbidden, HttpResponseRedirect
+from django.utils.translation import ugettext_lazy as _
+
+from .utils import compile_urls
+from ..models import AccessStateExtensionMixin as AccessState
+
+
+ALL_URLS = compile_urls([r'^'])
+
+
+class BasePermissionMiddleware:
+    """
+    Base class for middleware that allows or denies access to a resource.
+
+    This is a generic, fairly useless base class. It protects no pages and allows all
+    users through, although if by some miracle it does deny access to a user, it'll be
+    fine doing that.
+
+    Contains the following implemented methods:
+    - process_request: the method called on incoming request.
+    - deny_access: provides standard "you're not allowed" responses.
+
+    Contains the following hook methods:
+    - is_resource_protected: Returns True if and only if the middleware should be
+      protecting the resource the request is trying to access.
+    - deny_access_condition: Returns True if and only if the request should be disallowed
+      from accessing the protected resource.
+    - get_unauthorised_redirect_url: Returns the URL to redirect a denied GET request to
+      (default implementation returns /).
+    - get_access_denied_message: Returns the message to display when that happens
+      (default implementation returns '').
+    """
+    def is_resource_protected(self, request, **kwargs):
+        """
+        Hook for determining if a resource should be protected.
+
+        Returns True if and only if the middleware should be protecting
+        the page or endpoint the request is trying to access.
+        """
+        return False
+
+    def deny_access_condition(self, request, **kwargs):
+        """
+        Hook for disallowing access to a protected resource.
+
+        Returns True if and only if the request should be disallowed
+        from accessing the protected URL/page/endpoint.
+        """
+        return False
+
+    def get_unauthorised_redirect_url(self):
+        return '/'
+
+    def get_access_denied_message(self):
+        return ''
+
+    def deny_access(self, request, **kwargs):
+        """
+        Standard failure behaviour.
+
+        Returns HTTP 403 (Forbidden) for non-GET requests.
+
+        For GET requests, returns HTTP 302 (Redirect) pointing at either a URL specified
+        in the class's unauthorised_redirect attribute, if one exists, or / if not. This
+        version also adds a (translated) message if one is passed in.
+        """
+        # Raise a 403 for POST/DELETE etc.
+        if request.method != 'GET':
+            return HttpResponseForbidden()
+
+        # Add a message, if one has been defined.
+        message = self.get_access_denied_message()
+        if message:
+            messages.info(request, _(message))
```

Comment:
Oh sorry @meshy, didn't see your comment (thanks, Github autorefresh...).  Won't it be the same calculated value, though?  The first instance of `_` will cause `makemessages` to put it into the `.po` file, and the second will just do nothing because it's translating Spanish to Spanish or whatever/